Add completion and tweak behavior for fullpath

tyru opened the pull request:
Hello @pavoljuhas

Thanks for the great plugin!
I had been looking for a history plugin like this because it's so simple.
No need to save histories to separate files.

I've added (full) filepath completion to `:OldFileSearch`.
This pull request allows `<Tab>` key to work.

```
:OldFileSearch <Tab>
```

With that change, I had no choice but to change the behavior
**only when a fullpath was given**; skip the following process.

> (2) At least one pattern must match the tail component of the path.

tyru committed:
Add completion and tweak behavior for fullpath

* Add (full) filepath completion to :OldFileSearch

* If a fullpath was given, skip the following process.

> (2) At least one pattern must match the tail component of the path.

pavoljuhas committed:
Merge branch 'master' into tyru-completion.

Incorporate the fix of swapped smart case comparison.

pavoljuhas committed:
Restore requirement of a match in tail component.

Allow tail match also for patterns that contain path separators,
for which the pattern needs to span over the tail path.
`:OldFileSearch bin/` now finds old files in the bin directory,
but not in bin/backup/.

pavoljuhas committed:
Use 'nomagic' patterns when searching old files.

`:OldFileSearch .c` now matches .c files rather than files
containing the letter "c".

pavoljuhas commented:
Hi @tyru, Thank you for your suggestion, adding a `<Tab>` completion would be nice.

I think it is better to keep the original behavior and require some match in the tail, because in your version `:OldFileSearch / foo` may produce more matches that `:OldFileSearch foo`.  It is more consistent to see fewer matches after providing more patterns. I have updated the sources in my [completion branch](/pavoljuhas/oldfilesearch.vim/tree/completion) to support patterns containing file separator, so a single argument `in/old` should now find `plugin/oldfilesearch.vim`.

The Tab key behavior is not quite correct yet, because completion function uses only the current word, i.e.,

```
:OldFileSearch plugin o<Tab>
```

displays many more matches than

```
:OldFileSearch plugin o<Enter>
```

Can you update this PR so that completion function takes into account all patterns on the `:OldFileSearch` command line?

pavoljuhas commented:
PS: Please merge-in my completion branch before making any new commits.

tyru commented:
> Can you update this PR so that completion function takes into account all patterns on the :OldFileSearch command line?

Hmm, this seems impossible.
Because `<Tab>` can delete only a previous word but cannot delete other previous words.

```
:OldFileSearch plugin o<Tab>
```

in above example, `<Tab>` can delete 'o' and insert an arbitrary string,
but 'plugin' cannot be deleted.

pavoljuhas commented:
There is no need to delete "plugin" in that completion.  If "o" completes to a full path (which needs to match both "o" and "plugin") the entered command would then match only that path.

tyru commented:
@pavoljuhas Okay, I understand `:OldFileSearch` behavior.

But in the case of `<Tab>` behavior, I want to suggest that `<Tab>` doesn't require tail match.
Because when a user hits `<Tab>` key, a user may want to **confirm** current matching candidates
not only to **insert** (e.g. `list` value of `wildmode` option).

for example, `v:oldfiles` contains the following items:
- /home/tyru/.gitconfig
- /home/tyru/proj/.gitignore

And I hit `tyru<Tab>` to list the above items to confirm matching candidates.

---

But to implement the above behavior, I need to split `s:GetOldFiles()` function to some functions.
because `s:GetOldFiles()` returns a list filter()ed by the following conditions:
1. (1) All patterns must match the full path.
2. (2) At least one pattern must match the tail component of the path.
3. (3) Discard non-existing files.

So I want to move (2) condition to a dedicated function (like `s:FilterTailMatches()`).
And `s:OldFileComplete()` does not call `s:FilterTailMatches()` but `s:OldFileSearch()` does.
Would you like that?

tyru committed:
Implement <Tab> key behavior

1. It completes the candidates only matching to whole command-line arguments.
2. It completes all candidates when previous word(a:arglead) is empty.
  * Just see a:cmdline instead of a:arglead
3. It skips tail match.

tyru commented:
Implement <Tab> key behavior (quote from commit message).

> 1. It completes the candidates only matching to whole command-line arguments.
> 2. It completes all candidates when previous word(a:arglead) is empty.
>    - Just see a:cmdline instead of a:arglead
> 3. It skips tail match.

pavoljuhas commented:
> for example, v:oldfiles contains the following items:
> 
> /home/tyru/.gitconfig
> /home/tyru/proj/.gitignore
> 
> And I hit `tyru<Tab>` to list the above items to confirm matching candidates

I don't think such behavior is more useful, because it expands to _all_ oldfiles under your home directory.  It is more sensible to have `.git<Tab>` to complete to some `.gitignore` or `.gitattributes`, but NOT to `.git/info/exclude`.   I prefer to keep the plugin simple and consistent in its operation, so I want to keep the tail matching as is.

Please try again with a minimum change from my version that just considers the whole command line for producing completion matches.

With tail-matching active you can still catch some gitignore under your home directory with `tyru .g<Tab>`.  To complete _any_ oldfile under your home, you can do `tyru/ /<Tab>`, `tyru/ $<Tab>`, or `tyru/ \.<Tab>` - just pick whichever always-matching regular expression you prefer.

tyru commented:
Okay.

> Please try again with a minimum change from my version that just considers the whole command line for producing completion matches.

You mean you don't like the behavior both 2 and 3 of the followings?

> 1. It completes the candidates only matching to whole command-line arguments.
> 2. It completes all candidates when previous word(a:arglead) is empty.
> 3. It skips tail match.

pavoljuhas committed:
Simplify smart-case matching.

Apply smart-case matching per each pattern.  Use the \c, \C pattern
modifiers instead of case-specific comparison operators.

pavoljuhas committed:
Fix completion for OldFileComplete.

Use all command line arguments to generate completion choices.
If there are no arguments, fall back to empty pattern, which
matches everything.

pavoljuhas commented:
Yes, that's correct, see 96b594b6af249fd0a57e83b9b90d5658091efa40...58eb060c575f1149b2d1e9af0248f473ca8e621f for the behavior I had in mind.

Do you see some issues with that version?

tyru commented:
Okay, I will implement.
but i'm busy now, maybe on this weekend or later... sorry.

tyru committed:
Revert "Implement <Tab> key behavior"

This reverts commit 182c7ef8c620bd60e7b0b759868d5aae75f42eba.

tyru committed:
Merge remote-tracking branch 'origin/completion' into completion

tyru commented:
@pavoljuhas ah, I understand.
You have implemented `<Tab>` key behavior. thanks!

It looks good to me.
now I merged & pushed to my branch.
Could you merge this?

pavoljuhas commented:
Closed by d2d2cb048d2e459e1a9d0429d2452a434866558f.  Same content as in 3cb89c7.

tyru commented:
Thanks!
